+nosnoc/+discrete_time_problem/Gcs.m

```diff
@@ -0,0 +1,870 @@
+                            sigma_c_B = sigma_c_B + dcs.c_fun(x_ijk,z_ijk, v_global, p);
+                            sigma_lambda_B = sigma_lambda_B + obj.w.lambda(ii,jj-1,kk);
+                        end
+                        sigma_c_F = 0;
+                        sigma_lambda_F = 0;
+                        for kk=1:(opts.n_s + rbp)
+                            x_ijk = obj.w.x(ii,jj,kk);
+                            z_ijk = obj.w.z(ii,jj,kk);
+
+                            sigma_c_F = sigma_c_F + dcs.c_fun(x_ijk,z_ijk, v_global, p);
+                            sigma_lambda_F = sigma_lambda_F + obj.w.lambda(ii,jj,kk);
+                        end
+
+                        pi_c = sigma_c_B .* sigma_c_F;
+                        pi_lam = sigma_lambda_B .* sigma_lambda_F;
+                        nu = pi_c + pi_lam;
+                        eta = 1;
+                        for jjj=1:length(nu)
+                            eta = eta*nu(jjj);
+                        end
+                        eta_vec = [eta_vec;eta];
+                        delta_h = obj.w.h(ii,jj) - obj.w.h(ii,jj-1);
+                        obj.f = obj.f + obj.p.rho_h_p() * tanh(eta/opts.step_equilibration_sigma) * delta_h.^2;
+                    end
+                                end
+              case StepEquilibrationMode.l2_relaxed
+                eta_vec = [];
+                for ii=1:opts.N_stages
+                    p_stage = obj.p.p_time_var(ii);
+                    p =[p_global;p_stage];
+                    for jj=2:opts.N_finite_elements(ii)
+                        sigma_c_B = 0;
+                        sigma_lambda_B = 0;
+                        for kk=1:(opts.n_s + rbp)
+                            sigma_c_B = sigma_c_B + dcs.c_fun(x_ijk,z_ijk, v_global, p);
+                            sigma_lambda_B = sigma_lambda_B + obj.w.lambda(ii,jj-1,kk);
+                        end
+                        sigma_c_F = 0;
+                        sigma_lambda_F = 0;
+                        for kk=1:(opts.n_s + rbp)
+                            x_ijk = obj.w.x(ii,jj,kk);
+                            z_ijk = obj.w.z(ii,jj,kk);
+
+                            sigma_c_F = sigma_c_F + dcs.c_fun(x_ijk,z_ijk, v_global, p);
+                            sigma_lambda_F = sigma_lambda_F + obj.w.lambda(ii,jj,kk);
+                        end
+
+                        pi_c = sigma_c_B .* sigma_c_F;
+                        pi_lam = sigma_lambda_B .* sigma_lambda_F;
+                        nu = pi_c + pi_lam;
+                        eta = 1;
+                        for jjj=1:length(nu)
+                            eta = eta*nu(jjj);
+                        end
+                        eta_vec = [eta_vec;eta];
+                        delta_h = obj.w.h(ii,jj) - obj.w.h(ii,jj-1);
+                        obj.f = obj.f + obj.p.rho_h_p() * eta * delta_h.^2
+                    end
+                end
+              case StepEquilibrationMode.direct
```

Comment:
I am not sure does this option make sense at all.
the cross comps are also complementarity constraints. and we should let them be treated by the mpcc solver.
the only things we distinguish are they 
* in the objective (then does the indicator function also influce the weight, or is jus the rho_h, second branching, is it mean or diff). 
* hard constraints

Reply:
Yea I agree it may make sense to re-think exactly how the step equilibration is done. There are several ways to slice the existing modes. However I think this is probably a separate PR. As an aside the other mode now work for GCS too, them not working was due to a rebasing bug.

Reply:
Hence, lets make another issue :D 
At least the new issues are all quite small things (which will make review easy :D)

Reply:
Added issue: #111